includes/class-wc-payments-token-service.php

```diff
@@ -133,46 +208,73 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
+				$token->delete();
+			}
 		}
 		add_action( 'woocommerce_payment_token_deleted', [ $this, 'woocommerce_payment_token_deleted' ], 10, 2 );
 
 		return $tokens;
 	}
 
 	/**
-	 * Delete token from Stripe.
+	 * Clears unavailable tokens from the token list.
 	 *
-	 * @param string           $token_id Token ID.
-	 * @param WC_Payment_Token $token    Token object.
+	 * @param array  $tokens      Token list.
+	 * @param string $customer_id Customer ID.
+	 *
+	 * @return array Token list with tokens for $customer_id.
 	 */
-	public function woocommerce_payment_token_deleted( $token_id, $token ) {
-		if ( WC_Payment_Gateway_WCPay::GATEWAY_ID === $token->get_gateway_id() ) {
-			try {
-				$this->payments_api_client->detach_payment_method( $token->get_token() );
-				// Clear cached payment methods.
-				$this->customer_service->clear_cached_payment_methods_for_user( $token->get_user_id() );
-			} catch ( Exception $e ) {
-				Logger::log( 'Error detaching payment method:' . $e->getMessage() );
+	private function remove_unavailable_tokens( $tokens, $customer_id ) {
+		return array_filter(
+			$tokens,
+			function ( $token ) use ( $customer_id ) {
+				return $token->get_meta( self::CUSTOMER_ID_META_KEY ) === $customer_id;
 			}
-		}
+		);
 	}
 
 	/**
-	 * Set as default in Stripe.
+	 * Adds customer information to tokens that have none.
 	 *
-	 * @param string           $token_id Token ID.
-	 * @param WC_Payment_Token $token    Token object.
+	 * @param array $tokens  Token list.
+	 * @param int   $user_id WC user ID.
+	 *
+	 * @return array Token list with tokens for $customer_id.
 	 */
-	public function woocommerce_payment_token_set_default( $token_id, $token ) {
-		if ( WC_Payment_Gateway_WCPay::GATEWAY_ID === $token->get_gateway_id() ) {
-			$customer_id = $this->customer_service->get_customer_id_by_user_id( $token->get_user_id() );
-			if ( $customer_id ) {
-				$this->customer_service->set_default_payment_method_for_customer( $customer_id, $token->get_token() );
-				// Clear cached payment methods.
-				$this->customer_service->clear_cached_payment_methods_for_user( $token->get_user_id() );
+	private function migrate_customerless_tokens( $tokens, $user_id ) {
+		// If is_live cannot be determined, default it to true to avoid considering a live account as test.
+		$account_is_live = null === $this->account->get_is_live() || $this->account->get_is_live();
+		$customer_id     = $this->customer_service->get_customer_id_by_user_id(
+			$user_id,
+			$account_is_live ? 'live' : 'test'
+		);
+
+		// Fall back to current customer ID if the optimistic migration customer does not exist.
+		if ( ! $customer_id && $account_is_live ) {
+			$customer_id = $this->customer_service->get_customer_id_by_user_id( $user_id, 'test' );
+		}
+
+		if ( ! $customer_id ) {
+			return $tokens;
+		}
+
+		foreach ( $tokens as $token ) {
+			if ( ! $token->meta_exists( self::CUSTOMER_ID_META_KEY ) ) {
```

Comment:
As I understand `$tokens` might have other gateway tokens as well (e.g. Stripe). Would it make sense to check if `$tokent->get_gateway_id() === WC_Payment_Gateway_WCPay::GATEWAY_ID`? We do this [here](https://github.com/Automattic/woocommerce-payments/blob/c2d42e9bb505eb8498c5647f85a9f4090041e2c7/includes/class-wc-payments-token-service.php#L188-L188).

Reply:
Good catch, it does! Fixed in https://github.com/Automattic/woocommerce-payments/pull/973/commits/a625b21f42fb1799af0eef27626f5dbb7e4eab91.